File: leetcode/1007-minimum-domino-rotations-for-equal-row/solution.py
from typing import List


class Solution:
    def minDominoRotations(self, tops: List[int], bottoms: List[int]) -> int:
        def rotations_needed(target: int) -> int:
            rotate_top = rotate_bottom = 0
            for a, b in zip(tops, bottoms):
                if a != target and b != target:
                    return -1
                elif a != target:
                    rotate_top += 1
                elif b != target:
                    rotate_bottom += 1
            return min(rotate_top, rotate_bottom)

        result = rotations_needed(tops[0])
        if result != -1:
            return result
        return rotations_needed(bottoms[0])

File: leetcode/1007-minimum-domino-rotations-for-equal-row/Solution.java
class Solution {
    public int minDominoRotations(int[] tops, int[] bottoms) {
        int result = rotationsNeeded(tops, bottoms, tops[0]);
        if (result != -1) {
            return result;
        }
        return rotationsNeeded(tops, bottoms, bottoms[0]);
    }

    private int rotationsNeeded(int[] tops, int[] bottoms, int target) {
        int rotateTop = 0, rotateBottom = 0;
        for (int i = 0; i < tops.length; i++) {
            int a = tops[i], b = bottoms[i];
            if (a != target && b != target) {
                return -1;
            } else if (a != target) {
                rotateTop++;
            } else if (b != target) {
                rotateBottom++;
            }
        }
        return Math.min(rotateTop, rotateBottom);
    }
}
